python/ray/data/impl/plan.py

```diff
@@ -0,0 +1,134 @@
+from ray.data.block import Block
+from ray.data.impl.block_list import BlockList
+from ray.data.impl.compute import get_compute
+from ray.data.impl.stats import DatasetStats
+
+from typing import Callable, Tuple, Optional, Union, TYPE_CHECKING
+import uuid
+
+if TYPE_CHECKING:
+    import pyarrow
+
+
+class ExecutionPlan:
+    def __init__(self, in_blocks: BlockList, stats: DatasetStats):
+        self._in_blocks = in_blocks
+        self._out_blocks = None
+        self._in_stats = stats
+        self._out_stats = None
+        self._stages = []
+        stats.dataset_uuid = uuid.uuid4().hex
+
+    def with_stage(self, stage: "Stage"):
+        copy = ExecutionPlan(self._in_blocks, self._in_stats)
+        copy._stages = self._stages.copy()
+        copy._stages.append(stage)
+        return copy
+
+    def initial_num_blocks(self) -> int:
+        if self._out_blocks:
+            return self._out_blocks.initial_num_blocks()
+        for stage in self._stages[::-1]:
+            if stage.num_blocks is not None:
+                return stage.num_blocks
+        return self._in_blocks.initial_num_blocks()
+
+    def schema(
+        self, fetch_if_missing: bool = False
+    ) -> Union[type, "pyarrow.lib.Schema"]:
+        if self._stages:
+            if fetch_if_missing:
+                self.execute()
+            blocks = self._out_blocks
+        else:
+            blocks = self._in_blocks
+        metadata = blocks.get_metadata() if blocks else []
+        # Some blocks could be empty, in which case we cannot get their schema.
+        # TODO(ekl) validate schema is the same across different blocks.
+        for m in metadata:
+            if m.schema is not None:
+                return m.schema
+        if not fetch_if_missing:
+            return None
+        # Need to synchronously fetch schema.
+        return blocks.ensure_schema_for_first_block()
+
+    def meta_count(self) -> Optional[int]:
+        if self._stages:
+            blocks = self._out_blocks
+        else:
+            blocks = self._in_blocks
+        metadata = blocks.get_metadata() if blocks else None
+        if metadata and metadata[0].num_rows is not None:
+            return sum(m.num_rows for m in metadata)
+        else:
+            return None
+
+    def execute(self) -> BlockList:
+        # TODO: add optimizations:
+        # 1. task fusion
+        # 2. "move" block references into op()
+        if self._out_blocks is None:
+            blocks = self._in_blocks
```

Comment:
Note that these input blocks won't be released until `ExecutionPlan` is deallocated, which currently isn't done even after the output blocks have been computed. We might want to unset `self._in_blocks`, assuming that we are ok with not supporting multiple executions.

Reply:
Yeah, we'd have to push that down to the read task layer. Let me add that as a TODO.

Reply:
I have a followup that implements fusion, let's wait for that.

Reply:
Are you sure? Without this change, this will needlessly hold on to the input blocks, not only after the first stage is run, but even after the output blocks has been computed. This isn't a fusion consideration, we're holding on to an unnecessary block reference here.

Reply:
Synced offline, waiting for the follow-up in order to avoid merge conflicts.